EQL: Minimise CCS roundtrips

bpintea opened the pull request:
This introduces an optimisation of the EQL requests when these target
one remote cluster only (i.e. no mixed local and remote indices or
multiple remote clusters). In this case, the EQL request is forwarded
to the remote cluster and executed there, instead of having the local
cluster perform multiple queries of the remote cluster.

This behaviour is enabled by default, a new request parameter
`ccs_minimize_roundtrips` can be set to `false` to disable it.

bpintea committed:
Minimize CCS round-trips

This intorduces an optimisation of the EQL requests when these target
one remote cluster only (i.e. no mixed local and remote indices or
multiple remote clusters). In this case, the EQL request is forwarded
to the remote cluster and executed there, instead of having the local
cluster perform multiple querries of the remote cluster.

jrodewig commented on docs/reference/eql/eql-search-api.asciidoc:
```diff
@@ -89,6 +89,17 @@ request that targets only `bar*` still returns an error.
 Defaults to `true`.
 
+`ccs_minimize_roundtrips`::
+(Optional, Boolean)
++
+If `true`, network round-trips between the local and the remote cluster are
+minimized when executing cross-cluster search (CCS) requests.
```
Just fixes some wrapping.
```suggestion
(Optional, Boolean) If `true`, network round-trips between the local and the
remote cluster are minimized when running cross-cluster search (CCS) requests.
```

bpintea replied:
Thanks, fixed!

jrodewig commented on docs/reference/eql/eql-search-api.asciidoc:
```diff
@@ -89,6 +89,17 @@ request that targets only `bar*` still returns an error.
+If `true`, network round-trips between the local and the remote cluster are
+minimized when executing cross-cluster search (CCS) requests.
++
+This setting is currently only effective if the request does not include both
```
I'd use `option` or `parameter` to avoid confusion with cluster/index settings.
```suggestion
This option is currently only effective if the request does not include both
```

jrodewig reviewed (approved):
Docs LGTM. I left some non-blocking nits.

bpintea committed:
Adress doc review comments. Update error messages

Adress doc review comments. Update error messages

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

Luegg reviewed (approved):
Only some minor comments and questions, otherwise all makes sense to me 👍

jrodewig commented on docs/reference/eql/eql-search-api.asciidoc:
```diff
@@ -89,6 +89,15 @@ request that targets only `bar*` still returns an error.
 +
 Defaults to `true`.
 
+`ccs_minimize_roundtrips`::
```
We may want to link to https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-cross-cluster-search.html#ccs-network-delays.

Luegg replied:
👍 thanks a lot @jrodewig , I wasn't aware of this doc yet. That clarifies a lot!

bpintea replied:
Right, the setting follows suite of the one for search and is similar to it, but I've hesitated to add the reference, since we don't actually link the two. We could add an explanation note to clarify the distinction, though.

bpintea commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -155,6 +177,89 @@ static EqlSearchResponse createResponse(Results results, AsyncExecutionId id) {
+        if (clusterAliases.size() != 1 || clusterAliases.contains(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY)) {
+            return false;
+        }
+        // The remote async ID would not be valid on local cluster; furthermore on results fetching we would know neither if the ID is
```
It might, but wanted to have the `TODO` proposal reviewed, in case it emerges as a "blocker" for this optimisation.

bpintea commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -125,25 +133,39 @@ public static void operation(PlanExecutor planExecutor, EqlSearchTask task, EqlS
+            transportService.sendRequest(transportService.getRemoteClusterService().getConnection(clusterAlias),
+                EqlSearchAction.INSTANCE.name(), request.indices(remoteIndices), TransportRequestOptions.EMPTY,
+                new ActionListenerResponseHandler<>(wrap(r -> listener.onResponse(qualifyHits(r, clusterAlias)),
+                    e -> listener.onFailure(qualifyException(e, request.indices(), clusterAlias))),
```
No, it's a good point. I think current format was more of a copy that happened to work. Will update with next batch of changes.
And yes, we need the `remote-cluster-name:` prefix-stripped remote indices passed.

costin commented on x-pack/plugin/eql/qa/multi-cluster-with-security/src/test/java/org/elasticsearch/xpack/eql/RemoteClusterTestUtils.java:
```diff
@@ -15,11 +17,10 @@ public static String remoteClusterIndex(String indexName) {
     public static String remoteClusterPattern(String pattern) {
-        StringBuilder sb = new StringBuilder();
+        StringJoiner sj = new StringJoiner(",");
         for (String index: pattern.split(",")) {
```
For future reference `String.split` is expensive since it compiles a regex pattern - however it does have a shortcut (at least in the recent JVMs) where for single chars it does string iteration. For the former case better to use `Strings.commaDelimitedListToStringArray`

bpintea replied:
Thanks.

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -155,6 +177,89 @@ static EqlSearchResponse createResponse(Results results, AsyncExecutionId id) {
+            return false;
+        }
+        // The remote async ID would not be valid on local cluster; furthermore on results fetching we would know neither if the ID is
+        // remote or not, nor which remote cluster it belongs to (TODO: could rewrite the ID to smth like [alias:ID])
```
I'm okay with dealing with the async bit after this gets merged. Curious how async search and CCS deals with this scenario.

bpintea replied:
["Cross-cluster search is supported only with ccs_minimize_roundtrips set to `false`."](https://www.elastic.co/guide/en/elasticsearch/reference/current/async-search.html). I.e. it won't optimise it either, it'll query the remote cluster, but store the results locally.

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -155,6 +177,89 @@ static EqlSearchResponse createResponse(Results results, AsyncExecutionId id) {
+    }
+
+    // fixes the _index values by prefixing them with the source cluster alias' name
+    private static EqlSearchResponse qualifyHits(EqlSearchResponse r, String clusterAlias) {
+        EqlSearchResponse.Hits hits = r.hits();
+        List<EqlSearchResponse.Sequence> sequences = qualifySequences(hits.sequences(), clusterAlias);
+        List<EqlSearchResponse.Event> events = sequences != null ? null : qualifyEvents(hits.events(), clusterAlias, true);
+        return new EqlSearchResponse(new EqlSearchResponse.Hits(events, sequences, hits.totalHits()), r.took(), r.isTimeout());
+    }
+
+    private static List<EqlSearchResponse.Sequence> qualifySequences(List<EqlSearchResponse.Sequence> sequences, String clusterAlias) {
+        List<EqlSearchResponse.Sequence> qualifiedSequences = null;
+        if (sequences != null && sequences.isEmpty() == false) {
+            qualifiedSequences = new ArrayList<>(sequences.size());
+            for (EqlSearchResponse.Sequence s : sequences) {
+                qualifiedSequences.add(new EqlSearchResponse.Sequence(s.joinKeys(), qualifyEvents(s.events(), clusterAlias, false)));
+            }
+        }
+        return qualifiedSequences;
+    }
+
+    private static List<EqlSearchResponse.Event> qualifyEvents(List<EqlSearchResponse.Event> events, String clusterAlias,
+                                                               boolean emptyListIfNoEvents) {
+        List<EqlSearchResponse.Event> qualifiedEvents = null;
+        if (events != null && events.isEmpty() == false) {
+            qualifiedEvents = new ArrayList<>(events.size());
+            for (EqlSearchResponse.Event e : events) {
+                qualifiedEvents.add(new EqlSearchResponse.Event(buildRemoteIndexName(clusterAlias, e.index()), e.id(), e.source(),
+                    e.fetchFields()));
+            }
+        }
+        return qualifiedEvents == null && emptyListIfNoEvents ? new ArrayList<>() : qualifiedEvents;
+    }
```
Unless I'm missing something, these methods copy the response to prefix the index which is a waste.
Either change the actual returned objects to be mutable so that the remote cluster alias is added to the index or push the clusterAlias info so it's available during the `EqlSearchResponse.*` creation.

bpintea replied:
I've change the mutability to allow setting an index name. The alternative would require passing the cluster name thorough multiple layer, which is doable, but seemed a bit more contrived.

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -155,6 +177,89 @@ static EqlSearchResponse createResponse(Results results, AsyncExecutionId id) {
+                // index [...]"); this is also similar to a non-CCS `POST inexistent/_eql/search?ignore_unavailable=false`, but
+                // unfortunately unlike an inexistent pattern search: `POST inexistent*/_eql/search?ignore_unavailable=false, which raises a
+                // VerificationException as it's root cause. I.e. the failures are not homogenous.
+                finalException = new RemoteTransportException(e.getMessage(), new IndexNotFoundException(qualifiedIndex));
```
When @astefan come back please raise this with him since he worked on something similar. It's worth creating an issue to see whether the two errors can be aligned and if not why?

bpintea replied:
Great. It would seem good to me to return/report a not found index in the same way; i.e. no matter if the verifier finds the query error or other layer under EQL. (This could probably also be a subsequent PR.)

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/util/RemoteClusterRegistry.java:
```diff
@@ -39,10 +37,11 @@ public RemoteClusterRegistry(RemoteClusterService remoteClusterService, IndicesO
-            Strings.splitStringByCommaToArray(indexPattern));
-        indicesMap.remove(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY);
-        return indicesMap;
+    public Set<String> clusterAliases(String[] indices, boolean discardLocal) {
```
Does `discardLocal` has any other value than `true`? If not, better to keep the method specialized.

bpintea replied:
> Does `discardLocal` has any other value than `true`?

Yes, the optimisation code needs to know if the queried indices are all local or not (so it calls with this parameter set to `false`).

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -125,25 +133,39 @@ public static void operation(PlanExecutor planExecutor, EqlSearchTask task, EqlS
-        EqlConfiguration cfg = new EqlConfiguration(request.indices(), zoneId, username, clusterName, filter,
+        Set<String> clusterAliases = remoteClusterRegistry.clusterAliases(request.indices(), false);
+        if (canMinimizeRountrips(request, clusterAliases)) {
+            String clusterAlias = clusterAliases.iterator().next();
```
Instead of using two methods - `canMinimize` plus calling the cluster, how about combining them into a method that returns the cluster on which to re-route the call or otherwise returning null.
```
String singleCssCluster = minimizedCss(request, clusterAliases);
if (singleCssCluster != null) {
....
} else {
...
}
```

This helps with the different methods around checking, selecting and massaging the string into one method.

bpintea replied:
I did initially consider this format too (before the small RemoteClusterRegistry refactoring), but found delegating the decision to a method a clearer implementation (similar to [TransportSearchAction#shouldMinimizeRoundtrips()](https://github.com/elastic/elasticsearch/blob/4f22f437ee50cacb94b37b457be1da0b8ba0e8ce/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L332).

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -125,25 +133,39 @@ public static void operation(PlanExecutor planExecutor, EqlSearchTask task, EqlS
+            String clusterAlias = clusterAliases.iterator().next();
+            String[] remoteIndices = new String[request.indices().length];
+            for (int i = 0; i < request.indices().length; i++) {
+                remoteIndices[i] = request.indices()[i].substring(clusterAlias.length() + /*`:`*/1);
```
That inline comment is too l33t.  please move it after the statement or above it since as it stands requires multi-parsing.

bpintea replied:
Pulled that into a trailing comment.

costin reviewed (changes requested):
Left a first round of comments.
One thing that missing from this PR is the `client` EQL code which needs to be updated (in a backwards compatible way) with the new settings.

bpintea committed:
Adress review comments.

- rephrase docs;
- EqlSearchResponse.Event allows setting the index value;
- update comment.

Also, fix a case of index qualifying in an exception / failure case.

bpintea committed:
Merge branch 'master' into enh/optimize_ccs

bpintea commented on docs/reference/eql/eql-search-api.asciidoc:
```diff
@@ -89,6 +89,15 @@ request that targets only `bar*` still returns an error.
+(Optional, Boolean) If `true`, network round-trips between the local and the
+remote cluster are minimized when running cross-cluster search (CCS) requests.
++
+This option is currently only effective if the request does not include both
```
This explanation is compacter, though maybe less explicit. I've applied the proposal, though.

bpintea commented on x-pack/plugin/eql/qa/multi-cluster-with-security/src/test/java/org/elasticsearch/xpack/eql/EqlRestValidationIT.java:
```diff
@@ -17,21 +17,19 @@
     @Override
     protected String getInexistentIndexErrorMessage() {
-        return "\"caused_by\":{\"type\":\"verification_exception\",\"reason\":\"Found 1 problem\\nline -1:-1: Unknown index ";
+        return "\"root_cause\":[{\"type\":\"index_not_found_exception\",\"reason\":\"no such index ";
```
"Qualifying" the index in the exception replaces the VerificationException with an IndexNotFoundException, to ["harmonise"](https://github.com/elastic/elasticsearch/pull/76076/files#diff-88b74c40b99fe71512274733a696fcbf1cb53c34362b2b75892498d4d232cc0eR236) the failure reporting. This would be needed to not differentiate the failure format depending on which request parameters are set.

bpintea commented:
> One thing that missing from this PR is the client EQL code which needs to be updated (in a backwards compatible way) with the new settings.

[EqlClient](https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/main/java/org/elasticsearch/client/EqlClient.java)? I'll look into it.

costin commented:
> > One thing that missing from this PR is the client EQL code which needs to be updated (in a backwards compatible way) with the new settings.
> 
> [EqlClient](https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/main/java/org/elasticsearch/client/EqlClient.java)? I'll look into it.

Since it's a parameter, it's the request that needs to be changed not the client - e.g. [EqlSearchRequest](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchRequest.java)

bpintea committed:
Merge branch 'master' into enh/optimize_ccs

bpintea committed:
Address review comments

Add de-/serialization of ccs_minimize_roundtrips setting to
action/EqlSearchRequest.

bpintea committed:
Merge branch 'master' into enh/optimize_ccs

bpintea commented:
> One thing that missing from this PR is the client EQL code which needs to be updated (in a backwards compatible way) with the new settings.

Serialisation of the new parameter has been added.

astefan reviewed (approved):
LGTM

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  7.x  | Commit could not be cherrypicked due to conflicts |

To backport manually run `backport --upstream elastic/elasticsearch --pr 76076`
